server/src/main/java/com/defold/extender/services/DefoldSdkService.java

```diff
@@ -109,26 +96,45 @@ public boolean isLocalSdk(final String sdkVersion) {
         return sdkVersion == null || LOCAL_VERSION.equals(sdkVersion) || System.getenv("DYNAMO_HOME") != null;
     }
 
-    public File getSdk(String hash) throws IOException, URISyntaxException, ExtenderException {
-        return isLocalSdk(hash) ? getLocalSdk() : getRemoteSdk(hash);
+    public DefoldSdk getSdk(String hash) throws ExtenderException {
+        if (isLocalSdk(hash)){
+            return getLocalSdk();
+        }
+        // use ConcurrentHashMap with CompletableFuture here to avoid situation when
+        // several builds needs the same defoldsdk which doesn't exist locally. So one build job starts downloading,
+        // all other jobs wait for download complete and all of then continue running.
+        CompletableFuture<DefoldSdk> operation = operationCache.computeIfAbsent(hash, this::getRemoteSdk);
+        try (DefoldSdk sdk = operation.get()){
+            if (sdk == null) {
+                throw new ExtenderException(String.format("The given sdk does not exist: %s", hash));
+            }
+            if (!sdk.isValid()) {
+                throw new ExtenderException(String.format("The given sdk still does not exist: %s", hash));
+            }
+            evictCache();
+            LOGGER.info("Using Defold SDK version {}", hash);
+            return DefoldSdk.copyOf(sdk);
+        } catch (InterruptedException | ExecutionException e) {
+            LOGGER.error(String.format("The given sdk cannot be downloaded: %s", hash), e);
+            throw new ExtenderException(String.format("The given sdk cannot be downloaded: %s", hash));
+        } finally {
+            operationCache.remove(hash);
+        }
     }
 
-    public File getRemoteSdk(String hash) throws IOException, URISyntaxException, ExtenderException {
-        long methodStart = System.currentTimeMillis();
-
-        // Define SDK directory for this version
-        File sdkDirectory = new File(baseSdkDirectory.toFile(), hash);
-
-        // If directory does not exist, create it and download SDK
-        if (!Files.exists(sdkDirectory.toPath())) {
-            File lockFile = new File(baseSdkDirectory.toFile(), "tmp" + hash + ".lock");
-
-            if (lockFile.createNewFile()) { // atomic creation of lock file
-                try {
-                    boolean sdkFound = false;
-                    for (String url_pattern : REMOTE_SDK_URL_PATTERNS) {
-
-                        URL url = new URL(String.format(url_pattern, hash));
+    public CompletableFuture<DefoldSdk> getRemoteSdk(String hash) {
+        return CompletableFuture.supplyAsync(() -> {
+            long methodStart = System.currentTimeMillis();
+            // Define SDK directory for this version
+            File sdkDirectory = new File(baseSdkDirectory.toFile(), hash);
+            File sdkRootDirectory = new File(sdkDirectory, "defoldsdk");
+            DefoldSdk sdk = new DefoldSdk(sdkRootDirectory, hash, this);
```

Comment:
Create `DefoldSdk` object here to acquire reference before start downloading. It helps prevent issue when downloading will be completed but reference hasn't been acquired yet and other thread calls `evictCache` that leads to cleanup freshly downloaded sdk.